### PR TITLE
Plan(cannot add or edit tasks in plans)

### PR DESCRIPTION
**Pull request by Seris370**
Some conflicts are due to refactored MESSAGEs. Command 'lsplan' is changed to 'listplan'.
Didn't touch Ui part.

### CODE REVIEW

**Comment by jiayushe** [comment]
> Some conflicts are due to refactored MESSAGEs. Command 'lsplan' is changed to 'listplan'.
> Didn't touch Ui part.

Can you try to rebase your commits and resolve the conflicts manually?

This command should suffice:
`git pull --rebase https://github.com/AY1920S1-CS2103T-W11-1/main.git master`.

File: README.adoc [hunk 1]
```diff
@@ -14,7 +14,7 @@ ifndef::env-github[]
...
+Technical interviews are hard. Unless you are extraordinarily gifted, you will likely have to do tonnes of practice questions to ace them. And there are so many practice websites to choose from - Leetcode, Kattis and UVa, just to planName a few. Wouldn't it be nice if you could just consolidate all those questions into a single platform? Introducing AlgoBase (AB), a one stop desktop application designed for managing algorithm problems! AlgoBase can let you establish and modify your own training plan for algorithm problems flexibly and it can save several training plans at the same time on your local storage!  Also, you can use diverse tags and search to manage your questions and training plans conveniently. More importantly, AlgoBase is optimized for those who prefer to work with a Command Line Interface (CLI) while still having the benefits of a Graphical User Interface (GUI). If you can type fast, ALgoBase can get your problem management tasks done faster than traditional GUI apps.
```
Comment (jiayushe) [inline README.adoc]: Why are you changing `name` to `planName` in this line?

File: config/checkstyle/checkstyle.xml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!DOCTYPE module PUBLIC
-    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
-    "https://checkstyle.org/dtds/configuration_1_3.dtd">
+        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
```
Comment (jiayushe) [inline config/checkstyle/checkstyle.xml]: I don't think the changes in this file are relavant here, can you revert them?

**Review by jiayushe — CHANGES_REQUESTED** [review]
Please rebase your commits to current master. Also try to add some simple/placeholder tests if possible.

File: src/main/java/seedu/algobase/logic/commands/AddPlanCommand.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+public class AddPlanCommand extends Command {
+
+    public static final String COMMAND_WORD = "addplan";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Adds a Plan to the algobase. "
+            + "Parameters: "
+            + PREFIX_PLAN_NAME + "NAME "
+            + PREFIX_PLAN_DESCRIPTION + "DESCRIPTION "
+            + PREFIX_START_DATE + "START_DATE "
+            + PREFIX_END_DATE + "END_DATE "
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/AddPlanCommand.java]: Need a new line here.

File: src/main/java/seedu/algobase/logic/commands/EditPlanCommand.java [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    public static final String COMMAND_WORD = "editplan";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Edits the details of the Plan identified "
+            + "by the index number used in the displayed Plan list. "
+            + "Existing values will be overwritten by the input values.\n"
+            + "Parameters: INDEX (must be a positive integer) "
+            + "[" + PREFIX_PLAN_NAME + "NAME] "
+            + "[" + PREFIX_PLAN_DESCRIPTION + "DESCRIPTION] "
+            + "[" + PREFIX_START_DATE + "START_DATE] "
+            + "[" + PREFIX_END_DATE + "END_DATE] "
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/EditPlanCommand.java]: New line here.

File: src/main/java/seedu/algobase/logic/commands/FindPlanCommand.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import seedu.algobase.commons.core.Messages;
+import seedu.algobase.model.Model;
+import seedu.algobase.model.plan.NameContainsKeywordsPredicate;
+
+
+/**
+ * Finds and lists all plans in algobase whose name contains any of the argument keywords.
+ * Keyword matching is case insensitive.
+ */
+public class FindPlanCommand extends Command {
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/commands/FindPlanCommand.java:14]: I'll implement `findplan` that's similar to `find` later, which supports searching by more fields.

File: src/main/java/seedu/algobase/logic/parser/CliSyntax.java [hunk 6]
```diff
@@ -18,4 +18,8 @@
     public static final Prefix PREFIX_SORTING_ORDER = new Prefix("ord/");
     public static final Prefix PREFIX_PLAN = new Prefix("pl/");
     public static final Prefix PREFIX_PROBLEM = new Prefix("pr/");
+    public static final Prefix PREFIX_PLAN_NAME = new Prefix("n/");
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/parser/CliSyntax.java]: These are existing prefixes so no need to re-declare them.

File: src/main/java/seedu/algobase/logic/parser/ParserUtil.java [hunk 7]
```diff
@@ -55,6 +58,21 @@ public static Name parseName(String name) throws ParseException {
         return new Name(trimmedName);
     }
 
+    /**
+     * Parses a {@code String name} into a {@code PlanName}.
+     * Leading and trailing whitespaces will be trimmed.
+     *
+     * @throws ParseException if the given {@code name} is invalid.
```
Comment (le0tan) [inline src/main/java/seedu/algobase/logic/parser/ParserUtil.java:65]: I suppose it's not required to write thrown exception in JavaDoc for this project? (checkstyle doesn't complain on this anyways...)

**Review by le0tan — APPROVED** [review]
Closes #39

File: config/checkstyle/checkstyle.xml (hunk 2 above)
Comment (jiayushe) [reply config/checkstyle/checkstyle.xml]: Can you revert the changes in this file?

File: src/main/java/seedu/algobase/model/AlgoBase.java [hunk 8]
```diff
@@ -121,11 +151,18 @@ public String toString() {
...
     @Override
     public int hashCode() {
         return problems.hashCode();
     }
+
+    @Override
+    public ObservableList<Plan> getPlanList() {
+        return plans.asUnmodifiableObservableList();
+    }
+
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/AlgoBase.java]: This method should be under plan-level operations.

File: src/main/java/seedu/algobase/model/Model.java [hunk 9]
```diff
@@ -100,8 +134,4 @@
      */
     void updateSortedProblemList(Comparator<Problem> problemComparator);
 
-    //=========== Plan List =============================================================
-
-    /** Returns an unmodifiable view of the filtered Problem list */
-    ObservableList<Plan> getFilteredPlanList();
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/Model.java]: Move the plan-level methods here.

File: src/main/java/seedu/algobase/model/ModelManager.java [hunk 10]
```diff
@@ -147,19 +180,14 @@ public void updateSortedProblemList(Comparator<Problem> problemComparator) {
...
+    //============== Util ==================================================================
 
-    /**
-     * Returns an unmodifiable view of the list of {@code Problem} backed by the internal list of
-     * {@code versionedAlgoBase}
-     */
     @Override
-    public ObservableList<Plan> getFilteredPlanList() {
-        return filteredPlans;
+    public void updateFilteredPlanList(Predicate<Plan> predicate) {
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/ModelManager.java:185]: Why is this under `Util`?

File: src/main/java/seedu/algobase/model/plan/Plan.java [hunk 11]
```diff
@@ -10,41 +10,44 @@
 
 import seedu.algobase.model.task.Task;
 
+
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/plan/Plan.java]: Why new line here?

File: src/main/java/seedu/algobase/model/plan/Plan.java [hunk 12]
```diff
@@ -101,15 +104,15 @@ public boolean equals(Object other) {
...
     }
 
     @Override
     public String toString() {
         final StringBuilder builder = new StringBuilder();
-        builder.append(getName())
-                .append(" Description: ")
-                .append(getDescription())
+        builder.append(getPlanName())
+                .append(" PlanDescription: ")
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/plan/Plan.java]: Hmm this is for displaying purpose, so either `Description` or `Plan Description`?

File: src/main/java/seedu/algobase/model/plan/UniquePlanList.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+ * A Plan is considered unique by comparing using {@code Plan#isSamePlan(Plan)}. As such, adding and
+ * updating of plans uses Plan#isSamePlan(Plan) for equality so as to ensure that the Plan being added
+ * or updated is unique in terms of identity in the UniquePlanList. However, the removal of a Plan uses
+ * Plan#equals(Object) so as to ensure that the Plan with exactly the same fields will be removed.
+ *
+ * Supports a minimal set of list operations.
+ *
+ * @see Plan#isSamePlan(Plan)
+ */
+public class UniquePlanList implements Iterable<Plan> {
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/plan/UniquePlanList.java]: I don't think we need to enforce uniqueness on plans? @le0tan What do you think?

There is already an existing PlanList class anyway, so remove this class and modify the other one if necessary.

File: src/main/java/seedu/algobase/model/plan/exceptions/DuplicatePlanException.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
+package seedu.algobase.model.plan.exceptions;
+
+/**
+ * Signals that the operation will result in duplicate Plans (Plans are considered duplicates if they have the
+ * same identity).
+ */
+public class DuplicatePlanException extends RuntimeException {
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/plan/exceptions/DuplicatePlanException.java]: Same as above. I don't think we need uniqueness on plans.

File: src/test/java/seedu/algobase/logic/commands/DefaultModelStub.java [hunk 15]
```diff
@@ -84,11 +83,36 @@ public void setProblem(Problem target, Problem editedProblem) {
         throw new AssertionError("This method should not be called.");
     }
 
+    @Override
+    public void updateFilteredPlanList(Predicate<Plan> predicate) {
+
```
Comment (jiayushe) [inline src/test/java/seedu/algobase/logic/commands/DefaultModelStub.java]: Add `throw new AssertionError("This method should not be called.");` in the method body to be consistent with the rest.

**Review by jiayushe — CHANGES_REQUESTED** [review]

File: src/test/java/seedu/algobase/logic/commands/DefaultModelStub.java [hunk 16]
```diff
@@ -84,11 +83,36 @@ public void setProblem(Problem target, Problem editedProblem) {
...
+    }
+
     @Override
     public ObservableList<Problem> getFilteredProblemList() {
         throw new AssertionError("This method should not be called.");
     }
 
+    @Override
+    public boolean hasPlan(Plan plan) {
+        return false;
```
Comment (jiayushe) [inline src/test/java/seedu/algobase/logic/commands/DefaultModelStub.java]: Hmm should be `throw new AssertionError("This method should not be called.");` right?

**Review by jiayushe — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/algobase/logic/parser/AddPlanCommandParser.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+            description = ParserUtil.parsePlanDescription(argMultimap.getValue(PREFIX_DESCRIPTION).get());
+        } else {
+            description = PlanDescription.DEFAULT_PLAN_DESCRIPTION;
+        }
+
+        LocalDateTime startDate;
+        if (arePrefixesPresent(argMultimap, PREFIX_START_DATE)) {
+            startDate = ParserUtil.parseDate(argMultimap.getValue(PREFIX_START_DATE).get());
+        } else {
+            startDate = LocalDateTime.now();
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/logic/parser/AddPlanCommandParser.java:54]: Fix this in next PR.

**Review by jiayushe — APPROVED** [review]

**Pull request merged**